boa/src/profiler.rs

```diff
@@ -0,0 +1,93 @@
+#![allow(missing_copy_implementations, missing_debug_implementations)]
+
+#[cfg(feature = "profiler")]
+use measureme::{EventId, Profiler, TimingGuard};
+#[cfg(feature = "profiler")]
+use once_cell::sync::OnceCell;
+use std::fmt::{self, Debug};
+#[cfg(feature = "profiler")]
+use std::{
+    path::Path,
+    thread::{current, ThreadId},
+};
+
+/// MmapSerializatioSink is faster on macOS and Linux
+/// but FileSerializationSink is faster on Windows
+#[cfg(not(windows))]
+#[cfg(feature = "profiler")]
+type SerializationSink = measureme::MmapSerializationSink;
+#[cfg(windows)]
+#[cfg(feature = "profiler")]
+type SerializationSink = measureme::FileSerializationSink;
+#[cfg(feature = "profiler")]
+pub struct BoaProfiler {
+    profiler: Profiler<SerializationSink>,
+}
+
+#[cfg(feature = "profiler")]
+static mut INSTANCE: OnceCell<BoaProfiler> = OnceCell::new();
+
+#[cfg(feature = "profiler")]
+impl BoaProfiler {
+    pub fn start_event(&self, label: &str, category: &str) -> TimingGuard<'_, SerializationSink> {
+        let kind = self.profiler.alloc_string(category);
+        let id = EventId::from_label(self.profiler.alloc_string(label));
+        let thread_id = Self::thread_id_to_u32(current().id());
+        self.profiler
+            .start_recording_interval_event(kind, id, thread_id)
+    }
+
+    pub fn default() -> BoaProfiler {
+        let profiler = Profiler::new(Path::new("./my_trace")).unwrap();
+        BoaProfiler { profiler }
+    }
+
+    pub fn global() -> &'static BoaProfiler {
+        unsafe { INSTANCE.get_or_init(Self::default) }
+    }
+
+    pub fn drop(&self) {
+        // In order to drop the INSTANCE we need to get ownership of it, which isn't possible on a static unless you make it a mutable static
+        // mutating statics is unsafe, so we need to wrap it as so.
+        // This is actually safe though because init and drop are only called at the beginning and end of the application
+        unsafe {
+            INSTANCE
+                .take()
+                .expect("Could not take back profiler instance");
+        }
+    }
```

Comment:
I'm afraid this is not safe. If we call `forward()` more than once (which we sometimes do), this would be undefined behaviour.

I think the best solution is not to use a `static` global here. We could use a `Mutex` to avoid this, but I think this would reduce performance a lot. I think the best option is to conditionally have a `BoaProfiler` in the `Lexer`, `Parser` and `Interpreter`.

In the future, we could have a global `Boa` object with this field. In any case, I think the three of them should have a `start_event()` that would be no-op if the feature is not active. And all event starts should call to the `Lexer`, `Parser` or `Interpreter`.